include/envoy/filesystem/filesystem.h

```diff
@@ -4,68 +4,74 @@
 #include <memory>
 #include <string>
 
+#include "envoy/api/os_sys_calls.h"
 #include "envoy/common/pure.h"
-#include "envoy/event/dispatcher.h"
-#include "envoy/thread/thread.h"
 
 #include "absl/strings/string_view.h"
 
 namespace Envoy {
 namespace Filesystem {
 
 /**
- * Abstraction for a file on disk.
+ * Abstraction for a basic file on disk.
  */
 class File {
 public:
   virtual ~File() {}
 
   /**
-   * Write data to the file.
+   * Open the file with O_RDWR | O_APPEND | O_CREAT
+   * The file will be closed when this object is destructed
+   *
+   * @return bool whether the open succeeded
+   */
+  virtual Api::SysCallBoolResult open() PURE;
+
+  /**
+   * Write the buffer to the file. The file must be explicitly opened before writing.
+   *
+   * @return ssize_t number of bytes written, or -1 for failure
+   */
+  virtual Api::SysCallSizeResult write(absl::string_view buffer) PURE;
+
+  /**
+   * Close the file.
+   *
+   * @return bool whether the close succeeded
    */
-  virtual void write(absl::string_view) PURE;
+  virtual Api::SysCallBoolResult close() PURE;
 
   /**
-   * Reopen the file.
+   * @return bool is the file open
    */
-  virtual void reopen() PURE;
+  virtual bool isOpen() PURE;
 
   /**
-   * Synchronously flush all pending data to disk.
+   * @return string the file path
    */
-  virtual void flush() PURE;
+  virtual std::string path() PURE;
+
+  /**
+   * @return string a human-readable string describing the error code
+   * TODO(sesmith177) Use the IOError class after #5829 merges
```

Comment:
I'm going to merge this soon, so up to you if you want to merge that in or just do a quick follow up here.

Reply:
we've got a follow-up queued up; we will fix this there